src/raw/requests.rs

```diff
@@ -0,0 +1,587 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+use crate::{
+    kv_client::{HasError, KvClient, KvRawRequest, RpcFnType, Store},
+    pd::PdClient,
+    raw::ColumnFamily,
+    BoundRange, Error, Key, KvPair, Result, Value,
+};
+
+use futures::future::BoxFuture;
+use futures::prelude::*;
+use futures::stream::BoxStream;
+use kvproto::kvrpcpb;
+use kvproto::tikvpb::TikvClient;
+use std::mem;
+use std::sync::Arc;
+
+pub trait RawRequest: Sync + Send + 'static + Sized + Clone {
+    type Result;
+    type RpcRequest;
+    type RpcResponse: HasError + Clone + Send + 'static;
+    type KeyType;
+    const REQUEST_NAME: &'static str;
+    const RPC_FN: RpcFnType<Self::RpcRequest, Self::RpcResponse>;
+
+    fn execute(
+        mut self,
+        pd_client: Arc<impl PdClient>,
+    ) -> BoxFuture<'static, Result<Self::Result>> {
+        let stores = self.store_stream(pd_client);
+        Self::reduce(
+            stores
+                .and_then(move |(key, store)| {
+                    let request = self.clone().into_request(key, &store);
+                    store.dispatch::<Self>(&request, store.call_options())
+                })
+                .map_ok(move |r| Self::map_result(r))
+                .boxed(),
+        )
+    }
+
+    fn store_stream<PdC: PdClient>(
+        &mut self,
+        pd_client: Arc<PdC>,
+    ) -> BoxStream<'static, Result<(Self::KeyType, Store<PdC::KvClient>)>>;
+
+    fn into_request<KvC: KvClient>(
+        self,
+        key: Self::KeyType,
+        store: &Store<KvC>,
+    ) -> Self::RpcRequest;
+
+    fn map_result(result: Self::RpcResponse) -> Self::Result;
+
+    fn reduce(
+        results: BoxStream<'static, Result<Self::Result>>,
+    ) -> BoxFuture<'static, Result<Self::Result>>;
+}
+
+#[derive(Clone)]
+pub struct RawGet {
+    pub key: Key,
+    pub cf: Option<ColumnFamily>,
+}
+
+impl RawRequest for RawGet {
+    type Result = Option<Value>;
+    type RpcRequest = kvrpcpb::RawGetRequest;
+    type RpcResponse = kvrpcpb::RawGetResponse;
+    type KeyType = Key;
+    const REQUEST_NAME: &'static str = "raw_get";
+    const RPC_FN: RpcFnType<Self::RpcRequest, Self::RpcResponse> = &TikvClient::raw_get_async_opt;
```

Comment:
What's your consideration of using an associated const instead of a trait function?

Reply:
It would just be a redirection, so it seems like unnecessary boilerplate.